### PR TITLE
initial push of fieldmaps and test for fieldmaps

### PR DESCRIPTION
**Pull request by alexjaffray**
Here is my initial commit of the fieldmapping code and the test case (very simple for now but is along the right lines). 

Open to suggestions and discussion.

### CODE REVIEW

File: MRIFieldmaps/src/estimation.jl [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+"""
+R(x::Matrix{T})
+Regularization term which penalizes roughness in a first-order finite-difference sense
+# Arguments
+* `x::Matrix{T}` - fieldmap estimate (in radians)
+"""
+function R(x::Matrix{T}) where T
+
+    return T.(0.5) * sum(abs2,diff(x,dims=1),dims=[1,2]) + T.(0.5) * sum(abs2,diff(x,dims=2),dims=[1,2])
```
Comment (JeffFessler) [inline MRIFieldmaps/src/estimation.jl:25]:
```suggestion
    return T(0.5) * sum(abs2, diff(x,dims=1)) + T(0.5) * sum(abs2, diff(x,dims=2))
```
`sum` without `dims` arg will sum it all which is what we want, right?

Comment (alexjaffray) [reply MRIFieldmaps/src/estimation.jl:25]: extraneous `dims` also in `ml_cost`, will change.

File: MRIFieldmaps/src/estimation.jl [hunk 2]
```diff
@@ -0,0 +1,107 @@
+"""
+ml_cost(x::Matrix{T},y::Matrix{Complex{T}},z::Matrix{Complex{T}}, β)
```
[the remark this answers is not in the document]
Comment (alexjaffray) [reply MRIFieldmaps/src/estimation.jl:2]: Thanks, noted this detail for future docstrings.

File: MRIFieldmaps/src/estimation.jl (hunk 1 above)
Comment (alexjaffray) [reply MRIFieldmaps/src/estimation.jl]: fixed in latest commit

File: MRIFieldmaps/src/estimation.jl (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexjaffray) [reply MRIFieldmaps/src/estimation.jl]: fixed in latest commit

**Comment by tknopp** [comment]
@alexjaffray: I am closing this here, since we decided in https://github.com/MagneticResonanceImaging/MRIReco.jl/issues/132 to move it to another Github Repo https://github.com/MagneticResonanceImaging/MRIfieldmaps.jl. Could you reopen in that repository?

**Comment by alexjaffray** [comment]
Yes, I can do that.

Thanks,
Alex

**Comment by alexjaffray** [comment]
@JeffFessler Do you think it would be useful to move this code into the MRIFieldmaps.jl repo?

**Comment by JeffFessler** [comment]
Based on Figs. 4, 6, 9, 11 of [our 2020 paper](https://doi.org/10.1109/TCI.2020.3031082), I am fairly confident that the newer algorithms that I reproduced in MRIFieldmaps.jl will be faster than the QM method from our 2008 paper.  I didn't code up the 2008 in the new repo for two reasons: expecting it would be slower, and to avoid duplicating the work you did already on it.
Of course those 2020 results were in Matlab so it is possible (but unlikely) that the situation would be different in Julia.  Have you tried the new repo and does it produce similar fieldmaps for your data?  Is it faster?  It would be fine with me to have the 2008 method in the other repo even if it is mainly for historical reasons and to facilitate comparisons, so feel free to make a PR there.

**Comment by tknopp** [comment]
Yes, the fieldmap methods should go into MRIFieldmaps.jl and I would encourage you @alexjaffray to prepare a PR.

**Pull request closed without merging**